marlowe-playground-client/src/Marlowe/Parser.purs

```diff
@@ -49,17 +73,58 @@ instance prettyTerm :: Pretty a => Pretty (Term a) where
   prettyFragment (Term a) = prettyFragment a
   prettyFragment (Hole name _ _ _) = Leijen.text $ "?" <> name
 
-hole :: forall a. Parser String (Term a)
-hole = do
-  (ParseState _ start _) <- MonadState.get
-  name <- fromCharArray <$> (char '?' *> many alphaNum)
-  (ParseState _ end _) <- MonadState.get
-  pure $ Hole name Proxy start end
+-- a concrete type for holes only
+data MarloweHole = MarloweHole String MarloweType Position Position
 
-parseTerm :: forall a. Parser String a -> Parser String (Term a)
-parseTerm p = hole <|> (Term <$> p)
+class IsMarloweType a where
+  marloweType :: Proxy a -> MarloweType
+
+instance stringIsMarloweType :: IsMarloweType String where
+  marloweType _ = StringType
+
+instance bigIntegerIsMarloweType :: IsMarloweType BigInteger where
+  marloweType _ = BigIntegerType
+
+-- a Monoid for collecting Holes
+data Holes = Holes (Map String (Array MarloweHole))
+
+instance semigroupHoles :: Semigroup Holes where
+  append (Holes a) (Holes b) = Holes (Map.unionWith append a b)
+
+instance monoidHoles :: Monoid Holes where
+  mempty = Holes mempty
+
+{- | Find holes with the same name
+  We collect all the `Holes` using `getHoles` and then we fold through the result
+  to find occurances that have the same name as well as changing the values to be
+  single MarloweHoles rather than an array of MarloweHoles
+-}   
+validateHoles :: Holes -> Tuple (Array MarloweHole) (Map String MarloweHole)
+validateHoles (Holes m) = foldl f (Tuple [] mempty) arr
+  where
```

Comment:
I think you could probably neaten this up a bit with `foldlWithIndex`: https://pursuit.purescript.org/packages/purescript-foldable-traversable/4.1.1/docs/Data.FoldableWithIndex#v:foldrWithIndex

Reply:
yes, looks a bit cleaner